From 2958a0f2a5d06de99d58a8772205ec0d7f4629c0 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Fri, 20 Feb 2026 14:12:27 +0400
Subject: [PATCH] feat: make camera fly farther on gib + fix a bug

This is, as usual, compatible with vanilla, both servers and clients.

The bug that this fixes:
https://github.com/WofWca/quake3-better-gibs-mod/issues/3.

We also increase the default `cg_gibsExtraVerticalVelocity`
because with direct hit weapons the knockback is usually directed
a bit towards the ground, because in order to aim at the center
of a player your pitch needs to be a little below the horizon.
---
 code/cgame/cg_local.h |  1 +
 code/cgame/cg_main.c  |  4 +++-
 code/game/bg_pmove.c  | 35 +++++++++++++++++++++++++++++++++--
 code/game/bg_public.h |  8 ++++++++
 code/game/g_bot.c     |  5 +++++
 code/game/g_client.c  |  3 +++
 code/game/g_combat.c  | 13 +++++++++++++
 code/game/g_local.h   |  3 +++
 8 files changed, 69 insertions(+), 3 deletions(-)

diff --git a/code/cgame/cg_local.h b/code/cgame/cg_local.h
index b2f03803..90994f0d 100644
--- a/code/cgame/cg_local.h
+++ b/code/cgame/cg_local.h
@@ -1103,6 +1103,7 @@ extern	vmCvar_t		cg_gibsRandomVelocityFromKnockback;
 extern	vmCvar_t		cg_gibsExtraVerticalVelocity;
 extern	vmCvar_t		cg_gibsBounceFactor;
 extern	vmCvar_t		cg_gibsRotationFactor;
+extern	vmCvar_t		cg_gibsBetterCameraOnGib;
 extern	vmCvar_t		cg_drawTimer;
 extern	vmCvar_t		cg_drawFPS;
 extern	vmCvar_t		cg_drawSnapshot;
diff --git a/code/cgame/cg_main.c b/code/cgame/cg_main.c
index e8816489..5318ee6a 100644
--- a/code/cgame/cg_main.c
+++ b/code/cgame/cg_main.c
@@ -106,6 +106,7 @@ vmCvar_t	cg_gibsRandomVelocityFromKnockback;
 vmCvar_t	cg_gibsExtraVerticalVelocity;
 vmCvar_t	cg_gibsBounceFactor;
 vmCvar_t	cg_gibsRotationFactor;
+vmCvar_t	cg_gibsBetterCameraOnGib;
 vmCvar_t	cg_drawTimer;
 vmCvar_t	cg_drawFPS;
 vmCvar_t	cg_drawSnapshot;
@@ -229,9 +230,10 @@ static cvarTable_t cvarTable[] = {
 	{ &cg_gibsInheritPlayerVelocity, "cg_gibsInheritPlayerVelocity", "1.0", CVAR_ARCHIVE  },
 	{ &cg_gibsExtraRandomVelocity, "cg_gibsExtraRandomVelocity", "175", CVAR_ARCHIVE  },
 	{ &cg_gibsRandomVelocityFromKnockback, "cg_gibsRandomVelocityFromKnockback", "0.15", CVAR_ARCHIVE  },
-	{ &cg_gibsExtraVerticalVelocity, "cg_gibsExtraVerticalVelocity", "100", CVAR_ARCHIVE  },
+	{ &cg_gibsExtraVerticalVelocity, "cg_gibsExtraVerticalVelocity", "150", CVAR_ARCHIVE  },
 	{ &cg_gibsBounceFactor, "cg_gibsBounceFactor", "0.4", CVAR_ARCHIVE  },
 	{ &cg_gibsRotationFactor, "cg_gibsRotationFactor", "1.0", CVAR_ARCHIVE  },
+	{ &cg_gibsBetterCameraOnGib, "cg_gibsBetterCameraOnGib", "1", CVAR_USERINFO | CVAR_ARCHIVE  },
 	{ &cg_draw2D, "cg_draw2D", "1", CVAR_ARCHIVE  },
 	{ &cg_drawStatus, "cg_drawStatus", "1", CVAR_ARCHIVE  },
 	{ &cg_drawTimer, "cg_drawTimer", "0", CVAR_ARCHIVE  },
diff --git a/code/game/bg_pmove.c b/code/game/bg_pmove.c
index 820992cc..2a5f9a7b 100644
--- a/code/game/bg_pmove.c
+++ b/code/game/bg_pmove.c
@@ -1275,8 +1275,39 @@ static void PM_CheckDuck (void)
 
 	if (pm->ps->pm_type == PM_DEAD)
 	{
-		pm->maxs[2] = DEAD_MAXS_Z;
-		pm->ps->viewheight = DEAD_VIEWHEIGHT;
+		// Note that we want to support vanilla clients and servers,
+		// which do not have this `if` (they always run the `else` branch).
+		// On the server we check `cg_gibsBetterCameraOnGib` userinfo
+		// before setting `viewheight = NEW_GIBBED_VIEWHEIGHT`.
+		// Vanilla servers never set `viewheight` to `NEW_GIBBED_VIEWHEIGHT`,
+		// so on the client side this check is enough.
+		if ( pm->ps->viewheight == NEW_GIBBED_VIEWHEIGHT ) {
+			// Make the camera fly farther away when getting gibbed,
+			// by making the player's bounding box smaller
+			// and lifting it off the ground,
+			// as if the player became just their head.
+			//
+			// This also fixes the issue with gibs
+			// flying parallel to the ground even when knockback direction
+			// is towards the ground:
+			// https://github.com/WofWca/quake3-better-gibs-mod/issues/3.
+			//
+			// Note that we don't need to change `groundEntityNum` and stuff,
+			// because there is `PM_GroundTraceMissed()`
+			// right after `PM_CheckDuck()`.
+			pm->mins[2] = NEW_GIBBED_MINS_Z;
+			pm->maxs[2] = NEW_GIBBED_MAXS_Z;
+			// Also make them slimmer, because they're just a piece of meat now,
+			// and not an entire body.
+			pm->mins[0] = -PLAYER_WIDTH * 3 / 4;
+			pm->mins[1] = -PLAYER_WIDTH * 3 / 4;
+			pm->maxs[0] = PLAYER_WIDTH * 3 / 4;
+			pm->maxs[1] = PLAYER_WIDTH * 3 / 4;
+		} else {
+			pm->maxs[2] = DEAD_MAXS_Z;
+			pm->ps->viewheight = DEAD_VIEWHEIGHT;
+		}
+
 		return;
 	}
 
diff --git a/code/game/bg_public.h b/code/game/bg_public.h
index ef8b5a1b..52831868 100644
--- a/code/game/bg_public.h
+++ b/code/game/bg_public.h
@@ -55,6 +55,14 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 #define DEAD_MAXS_Z			-8
 #define	DEAD_VIEWHEIGHT		-16
 #define INVUL_RADIUS		42
+// Make sure to keep `mins[2]` negative so that player origin doesn't sink
+// into the ground, to guard against weird behavior
+// like sounds not being played (because they're "behind the wall")
+// or something.
+#define	NEW_GIBBED_MINS_Z		-2
+#define NEW_GIBBED_HEIGHT_DIFF	NEW_GIBBED_MINS_Z - MINS_Z
+#define	NEW_GIBBED_MAXS_Z		DEAD_MAXS_Z + NEW_GIBBED_HEIGHT_DIFF
+#define	NEW_GIBBED_VIEWHEIGHT	DEAD_VIEWHEIGHT + NEW_GIBBED_HEIGHT_DIFF
 
 //
 // config strings are a general means of communicating variable length strings
diff --git a/code/game/g_bot.c b/code/game/g_bot.c
index b6b205f1..544f96e9 100644
--- a/code/game/g_bot.c
+++ b/code/game/g_bot.c
@@ -727,6 +727,11 @@ static void G_AddBot( const char *name, float skill, const char *team, int delay
 	// don't send tinfo to bots, they don't parse it
 	Info_SetValueForKey( userinfo, "teamoverlay", "0" );
 
+	// Settings this to 1 for bots because this also fixes a bug
+	// https://github.com/WofWca/quake3-better-gibs-mod/issues/3.
+	// See `bg_pmove.c`.
+	Info_SetValueForKey( userinfo, "cg_gibsBetterCameraOnGib", "1" );
+
 	// register the userinfo
 	trap_SetUserinfo( clientNum, userinfo );
 
diff --git a/code/game/g_client.c b/code/game/g_client.c
index c6a0e874..5b4b2717 100644
--- a/code/game/g_client.c
+++ b/code/game/g_client.c
@@ -734,6 +734,9 @@ void ClientUserinfoChanged( int clientNum ) {
 		client->pers.predictItemPickup = qtrue;
 	}
 
+	client->pers.cg_gibsBetterCameraOnGib =
+		atoi( Info_ValueForKey( userinfo, "cg_gibsBetterCameraOnGib" ) );
+
 	// set name
 	Q_strncpyz ( oldname, client->pers.netname, sizeof( oldname ) );
 	s = Info_ValueForKey (userinfo, "name");
diff --git a/code/game/g_combat.c b/code/game/g_combat.c
index 3ff70bb1..48b55392 100644
--- a/code/game/g_combat.c
+++ b/code/game/g_combat.c
@@ -313,6 +313,18 @@ void GibEntity( gentity_t *self, int killer, const int damageBloodFallback ) {
 	self->takedamage = qfalse;
 	self->s.eType = ET_INVISIBLE;
 	self->r.contents = 0;
+
+	// See `NEW_GIBBED_VIEWHEIGHT` references in `bg_pmove.c`.
+	//
+	// Note that we only do this when gibbed and not on any death,
+	// because otherwise, due to the change to `mins[2]`,
+	// the body would immediately visually fall under ground.
+	if ( self->client && self->client->pers.cg_gibsBetterCameraOnGib & 0x1
+		// If we already died, we don't want to change
+		// the camera position again.
+		&& self->client->deathTime == level.time ) {
+		self->client->ps.viewheight = NEW_GIBBED_VIEWHEIGHT;
+	}
 }
 
 /*
@@ -523,6 +535,7 @@ void player_die( gentity_t *self, gentity_t *inflictor, gentity_t *attacker, int
 	}
 #endif
 	self->client->ps.pm_type = PM_DEAD;
+	self->client->deathTime = level.time;
 
 	if ( attacker ) {
 		killer = attacker->s.number;
diff --git a/code/game/g_local.h b/code/game/g_local.h
index a09333a0..b60399da 100644
--- a/code/game/g_local.h
+++ b/code/game/g_local.h
@@ -241,6 +241,7 @@ typedef struct {
 	qboolean	initialSpawn;		// the first spawn should be at a cool location
 	qboolean	predictItemPickup;	// based on cg_predictItems userinfo
 	qboolean	pmoveFixed;			//
+	int			cg_gibsBetterCameraOnGib;
 	char		netname[MAX_NETNAME];
 	int			maxHealth;			// for handicapping
 	int			enterTime;			// level.time the client entered the game
@@ -302,6 +303,8 @@ struct gclient_s {
 
 	int			lastKillTime;		// for multiple kill rewards
 
+	int			deathTime;			// 0 if alive
+
 	qboolean	fireHeld;			// used for hook
 	gentity_t	*hook;				// grapple hook if out
 
-- 
2.43.0

